snapcraft/internal/sources/_local.py

```diff
@@ -25,13 +26,79 @@
 
 class Local(Base):
 
+    def __init__(self, *args, copy_function=file_utils.link_or_copy, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.source_abspath = os.path.abspath(self.source)
+        self.copy_function = copy_function
+
+        self._ignore = functools.partial(
+            _ignore, self.source_abspath, os.getcwd())
+
     def pull(self):
-        current_dir = os.getcwd()
-        source_abspath = os.path.abspath(self.source)
+        file_utils.link_or_copy_tree(
+            self.source_abspath, self.source_dir, ignore=self._ignore,
+            copy_function=self.copy_function)
+
+    def _check(self, target):
+        try:
+            target_mtime = os.lstat(target).st_mtime
+        except FileNotFoundError:
+            return False
+
+        self._updated_files = set()
+        self._updated_directories = set()
+
+        for (root, directories, files) in os.walk(self.source_abspath,
+                                                  topdown=True):
+            ignored = set(self._ignore(root, directories + files))
+            if ignored:
+                # Prune our search appropriately given an ignore list, i.e.
+                # don't walk into directories that are ignored.
+                directories[:] = [d for d in directories if d not in ignored]
+
+            for file_name in (set(files) - ignored):
+                path = os.path.join(root, file_name)
+                if os.lstat(path).st_mtime >= target_mtime:
+                    self._updated_files.add(os.path.relpath(path, self.source))
+
+            for directory in directories:
+                path = os.path.join(root, directory)
+                if os.lstat(path).st_mtime >= target_mtime:
+                    # Don't decend into this directory-- we'll just copy it
+                    # entirely.
+                    directories.remove(directory)
 
-        def ignore(directory, files):
-            if directory == source_abspath or \
-               directory == current_dir:
+                    # os.walk will include symlinks to directories here, but we
+                    # want to treat those as files
+                    relpath = os.path.relpath(path, self.source)
+                    if os.path.islink(path):
+                        self._updated_files.add(relpath)
+                    else:
+                        self._updated_directories.add(relpath)
+
+        return (
+            len(self._updated_files) > 0 or
+            len(self._updated_directories) > 0)
+
+    def _update(self):
+        # First, copy the directories
+        for directory in self._updated_directories:
```

Comment:
this feels a lot like the pluginhandlers "migration" code. Maybe we need some generalization there (in the future)

Reply:
There are definitely some similarities in that they both operate on directories and files, but the pluginhandler does some special stuff as well.